content/blog/2020-12-18-type-and-effect-system.md

```diff
@@ -0,0 +1,93 @@
+
+In order to perform side-effect-aware static analysis, we introduce the idea of a _type-and-effect system_, which extends the notion of types to encode side effect information. For example, the OCaml function `print_int` has type `int -> unit`, but in a type-and-effect system, the signature might look more like `print_int : (int -> unit !output)`. The following section explores one implementation of such a system, available [here](https://github.com/kq-li/type-and-effect-system).
+
+# Language
+The language designed for this project operates over four types -- `unit`, `int`, `bool`, and one-argument functions `t1 -> t2` -- and four effects -- `input`, `output`, `read`, and `write`. In addition, the language includes control flow statements such as conditionals and loops. Finally, the language supports higher-order and anonymous functions.
+
+The standard library for the language includes, along with standard unary and binary operations, four side-effect-causing functions:
+- `scan : (unit -> int !input)` reads an integer from standard in
+- `print : (int -> unit !output)` writes an integer to standard out
+- `load : (int -> int !read)` reads an integer from the specified integer memory location 
+- `store : int -> (int -> unit !write)` writes an integer to the specified integer memory location
+
+Below is a sample program, which takes in a number as input and computes its square:
+
+```
+x : int !input = scan ();
+y : int = mul x x;
+_ : unit !output = print y;
+```
+
+Notice the special assignment syntax: every expression must be assigned to a variable (with `_` as a dummy) and every effect of an assigned expression must be encoded exactly in the type annotation. These requirements, which are enforced by the typechecker, guard against user error: it is impossible to cause a side effect without knowing about it.
+
+All functions are values, and use the lambda-calculus-inspired syntax `\x : t . s` in which `x` is the argument, `t` is its type, and `s` is a statement to execute. In addition, there is a short form `\x : t . e` where the statement is replaced by an expression to be immediately returned.
+
+Below is a sample program demonstrating functions and their properties:
+
+```
+add5 : int -> int = add 5;
```

Comment:
Presumably `add` is defined somewhere else?

Reply:
yep, clarified in the stdlib section